dotnet/src/SemanticKernel.Abstractions/AI/FunctionChoiceBehaviors/RequiredFunctionChoiceBehavior.cs

```diff
@@ -0,0 +1,163 @@
+    /// </summary>
+    /// <remarks>
+    /// This should be greater than or equal to <see cref="MaximumAutoInvokeAttempts"/>.
+    /// Once this limit is reached, the functions will no longer be included in subsequent requests that are part of the user operation, e.g.
+    /// if this is 1, the first request will include the functions, but the subsequent response sending back the functions' result
+    /// will not include the functions for further use.
+    /// </remarks>
+    [JsonPropertyName("maximumUseAttempts")]
+    public int MaximumUseAttempts { get; init; } = 1;
+
+    /// <inheritdoc />
+    public override FunctionChoiceBehaviorConfiguration GetConfiguration(FunctionChoiceBehaviorContext context)
+    {
+        bool autoInvoke = this.MaximumAutoInvokeAttempts > 0;
+
+        // If auto-invocation is specified, we need a kernel to be able to invoke the functions.
+        // Lack of a kernel is fatal: we don't want to tell the model we can handle the functions
+        // and then fail to do so, so we fail before we get to that point. This is an error
+        // on the consumers behalf: if they specify auto-invocation with any functions, they must
+        // specify the kernel and the kernel must contain those functions.
+        if (autoInvoke && context.Kernel is null)
+        {
+            throw new KernelException("Auto-invocation for Required choice behavior is not supported when no kernel is provided.");
+        }
+
+        List<KernelFunction>? requiredFunctions = null;
+        bool allowAnyRequestedKernelFunction = false;
+
+        // Handle functions provided via constructor as function instances.
+        if (this._functions is { } functions && functions.Any())
+        {
+            // Make sure that every function can be found in the kernel.
+            if (autoInvoke)
+            {
+                foreach (var function in functions)
+                {
+                    if (!context.Kernel!.Plugins.TryGetFunction(function.PluginName, function.Name, out _))
+                    {
+                        throw new KernelException($"The specified function {function.PluginName}.{function.Name} is not available in the kernel.");
+                    }
+                }
+            }
+
+            requiredFunctions = functions.ToList();
+        }
+        // Handle functions provided via the 'Functions' property as function fully qualified names.
+        else if (this.Functions is { } functionFQNs && functionFQNs.Any())
+        {
+            requiredFunctions = [];
+
+            foreach (var functionFQN in functionFQNs)
+            {
+                // Make sure that every function can be found in the kernel.
+                var name = FunctionName.Parse(functionFQN, FunctionNameSeparator);
+
+                if (!context.Kernel!.Plugins.TryGetFunction(name.PluginName, name.Name, out var function))
```

Comment:
Why do we skip the autoInvoke check here but require it on line 112?

Reply:
If the function is specified as a function FQN, then it must be registered in the kernel regardless of the invocation type. How and where else could the connector obtain the function metadata to advertise?  
   
On the other hand, if a function is specified as an instance, the requirement of being registered in the kernel only applies to functions in auto-invocation mode, as the connector will need to look it up in the kernel to auto-invoke it. For manual mode, the connector delegates function resolution to the caller that registered the function for invocation and possesses its instance as such.

I've added the comment to make it clearer.